### DISCUSSION TITLE
Spurious solid-mechanics kernels on lower-dimensional meshes

### DISCUSSION THREAD

**Discussion by ttruster** [idaholab/moose#27335] [Q&A Modules: Solid mechanics]
(Apologies for long post as usual from me) I was reminded recently of an issue to raise:

For solid mechanics models built using the master action system and that contain lower dimensional meshes for mortar constraints, there is not an error-catch for making a model with no block restriction in the master action.

Namely, what happens is, solid mechanics kernels and materials (strains) get generated for the lower-d mesh also, basically behaving like truss or membrane elements. The calculation runs just fine (as long as the stress material is made without a block restriction; otherwise a bunch of error messages are printed for missing stresses on the lower-d blocks). However, the result is likely not what the user intended (physics will be wrong).

See the two attached files, where the only difference is that `block = 0` has been deleted from 3 lines.
[two_lower_d_examples.zip](https://github.com/idaholab/moose/files/14925996/two_lower_d_examples.zip)

I welcome thoughts on a prudent fix. My 2 suggestions for the case of mixed dimensional meshes and no block restriction are: 1) only apply master actions to the highest-dimensional blocks or 2) throw an error message from the master action. Basically, I don't think there is a use case where the user intentionally wants all properties assigned to higher and lower dimensional meshes without discrimination; they should be notified of what they are doing (to save the 2 hours of debugging that it took me to notice this).

@reverendbedford and @hugary1995 are welcome to chime in on this.

PS: I should have brought this up before the solid mechanics renaming was rolled out; apologies for being swamped and missing that.

**Comment by jiangwen84** [comment]
I actually just found this issue for my simulation which uses Mortar.  It took me a while to figure it out and it is annoying because the results are totally wrong without ang error messages or warning. 

I like your second approach better. Throwing an error and let user know the issue.  You first approach would work too,  but it will hide another setting from users.

**Comment by GiudGiud** [comment]
I think this will be useful outside of solid mechanics as well. I can put it in the PhysicsBase class and when solid mechanics leverages Physics for its main action, it will be fixed that way

**Reply by lindsayad** [reply]
This seems like a genuine issue, so I've opened https://github.com/idaholab/moose/issues/27385
